File: Assets/Samples/TSBehaviour/Script/Example2.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using Puerts;
using Puerts.TSLoader;
using UnityEngine.Serialization;
using UnityEngine.UI;

namespace TsSamples
{
    public delegate string DelegateFunc5(int val);

    public delegate GameObject DelegateFunc6(string name);
    public class Example2 : MonoBehaviour
    {
        private JsEnv _mJsEnv;
        [FormerlySerializedAs("ModuleName")] public string moduleName; //可配置加载的js模块 
        
        private void OnEnable()
        {
            if (_mJsEnv == null)
            {
                string path = "D:/PuerTs/TestPuerTs/TsProject/output/";
                _mJsEnv = new JsEnv(new DefaultLoader(path), 8080);
            }
            
            //_mJsEnv.UsingFunc<GameObject,string>();
            
            JSObject jsObj  = _mJsEnv.ExecuteModule(moduleName + ".mjs");
            Action func1 = jsObj.Get<Action>("func1");
            func1();
            
            Action<string> func2 = jsObj.Get<Action<string>>("func2");
            func2("C# Call func2");

            Action<GameObject> func3 = jsObj.Get<Action<GameObject>>("func3");
           // Func<GameObject,string> func3 = jsObj.Get<Func<GameObject, string>>("func3");
            GameObject cube = GameObject.Find("Cube");
            func3(cube);
            
            Action<MonoBehaviour> func4 = jsObj.Get<Action<MonoBehaviour>>("func4");
            // Func<GameObject,string> func3 = jsObj.Get<Func<GameObject, string>>("func3");
            func4(this);

            _mJsEnv.UsingFunc<int,string>();
            DelegateFunc5 func5 = jsObj.Get<DelegateFunc5>("func5");
            var str = func5(123);
            Debug.Log("return str="+str);
            
            _mJsEnv.UsingFunc<string,GameObject>();
            DelegateFunc6 func6 = jsObj.Get<DelegateFunc6>("func6");
            var obj = func6("Cube");
            Debug.Log("find obj="+obj);
            
            Application.runInBackground = true;
        }

        public Action action1 = null;
        void Update()
        {
            action1?.Invoke();
            if (_mJsEnv != null)
                _mJsEnv.Tick();
        }
    }

    public class Example2TestClass
    {
        public string str = "Example2TestClass";

        public void LogStr(string str)
        {
            Debug.Log(str);
        }

        public static int TestAdd(int a, int b)
        {
            return a + b;
        }
    }
}

File: TsProject/src/Example2.ts
console.log("Example2");

export function func1() : void
{
    console.log("ts:call func1");
}

export function func2(s:string):void
{
    console.log("ts:call func2,s="+s);
}

let go:CS.UnityEngine.GameObject;
export function func3(obj:CS.UnityEngine.GameObject):void
{
    console.log('ts:call func3,go='+obj.name);
    go = obj;
    setInterval(()=>{
        go.transform.Rotate(300*CS.UnityEngine.Time.deltaTime,0,0);
    });
}

export function func4(component:any):void
{
    console.log('ts:call func4,c='+component);
    let a:number = 1;
    component.action1 = ()=>
    {
        go.transform.localScale = new CS.UnityEngine.Vector3(a,a,a);
        a += CS.UnityEngine.Time.deltaTime;
        a = a >=3 ? 1 : a;
    }
}

export function func5(val:number):string
{
    console.log("ts:call func5="+val);
    return "from func5 return data";
}

export function func6(objName:string):CS.UnityEngine.GameObject
{
    let findGo:CS.UnityEngine.GameObject = CS.UnityEngine.GameObject.Find(objName);
    console.log("ts:call func6");
    return findGo;
}

function testCallCSharp()
{
    let test = new CS.TsSamples.Example2TestClass();
    console.log("test class str = " + test.str);
    test.LogStr("testCallCSharp");
    let res = CS.TsSamples.Example2TestClass.TestAdd(1,2);
    console.log("res="+res);
}

testCallCSharp();